app/controllers/executions_controller.rb

```diff
@@ -0,0 +1,36 @@
+class ExecutionsController < ApplicationController
+  skip_before_action :check_admin
+
+  def results
+    submission = Submission.find(params[:submission_id])
+
+    if submission.user_id != current_user.id && !current_user.admin?
+      head :forbidden
+      return
+    end
+
+    num_passed = 0
+    failed_execution_ids = []
+
+    executions = Execution.select(:test_id, :passed).where(submission_id: params[:submission_id])
+
+    executions.each do |execution|
+      if execution.passed?
```

Comment:
Found out you actually need:
```ruby
executions.select { |e| !e.passed? }.map { |e| e.test_id }
```
because map will replace the any `e.passed?` with `nil`